### PR TITLE
Bug Fix: Cannot add immediate email notifications for new threads in a community

### LINKED ISSUE #614: Cannot add immediate email notifications for new threads in a community
Slack: https://hicommonwealth.slack.com/archives/CT2302L7L/p1597950579000500

### LINKED ISSUE #617: Handle missing subscriptions for chains/communities
- Why are there no options to select here? Should be "New threads"

image

- What is "Notifications on Comments"?

image

### LINKED ISSUE #618: Handle indeterminate state for "new threads (all communities)"
The state displayed for "New threads (all communities)" here is inaccurate, since I am only subscribed to new threads in some communities

image

### PR DESCRIPTION
**Pull request by zakhap**
## Description

When diagnosing #614, I believe the issue to be that there is no subscription to update for the community `edgeware` in the bug video, thus it throws the error. This would mean any user with roles created prior to roles auto-subscribing would experience this bug. I wrote a migration that checks subscriptions and roles, creating a new subscription for each community of which there is a role without one.  This addresses #617 explicitly.

Merged in `zak.more-subscription-fixing` (previously PR #672) which attended to some overlapping issues with #617 as well as #618. 

~@dillchen I'd love a QA prior to code review to check and verify this fixes the bug you identified. If not... I've got more work ahead of me.~

Closes #614, as well as closes #617 and closes #618,

## Motivation and Context

🐛 🐞 

## How has this been tested?

Migration has been tested by running it.

## Have proper tags been added (for bug, enhancement, breaking change)?
- [x] yes

## Does this PR affect any server routes?
- [x] no

### CODE REVIEW

File: server/migrations/20200826183259-add-subscriptions-for-roles.js [hunk 1]
```diff
@@ -0,0 +1,64 @@
...
+        // Get all addresses for User
+        const resAddresses = await queryInterface.sequelize.query(`SELECT * FROM "Addresses" WHERE user_id=${id};`, { transaction: t, });
+        const addresses = resAddresses[0];
+        const userRoles = [];
+
+        await Promise.all(addresses.map(async (address) => {
+          // For each address, get roles and add them to userRoles
+          const addressRoles = await queryInterface.sequelize.query(`SELECT * FROM "Roles" WHERE address_id=${address.id};`, { transaction: t, });
+          addressRoles[0].forEach((r) => userRoles.push(r));
+        }));
```
Comment (raykyri) [inline server/migrations/20200826183259-add-subscriptions-for-roles.js:21]: This could all be one query

Comment (zakhap) [reply server/migrations/20200826183259-add-subscriptions-for-roles.js]: fixed this query

File: client/scripts/views/pages/subscriptions/notificationSettings.ts [hunk 2]
```diff
@@ -89,25 +89,22 @@ interface IBatchedSubscriptionRowAttrs {
...
     const someEmail = subscriptions.some((s) => s.immediateEmail);
-    if (everyActive && someEmail) {
+    const everyEmail = subscriptions.some((s) => s.immediateEmail);
+    if (everyActive && everyEmail) {
       vnode.state.option = 'Notifications on (app + email)';
-    } else if (everyActive) {
+    } else if (everyActive && !someEmail) {
       vnode.state.option = 'Notifications on (app only)';
+    } else if (someActive) {
+      vnode.state.option = 'Some notifications on';
```
Comment (raykyri) [inline client/scripts/views/pages/subscriptions/notificationSettings.ts:107]: The `New threads (all communities)` option should show "Some notifications on" if the new threads option is only selected for some communities, right?

Comment (zakhap) [reply client/scripts/views/pages/subscriptions/notificationSettings.ts:107]: Correct. It is the "indeterminate" state for the dropdown menu. It's not an option to select, but a default value for the button trigger to the menu.

**Pull request merged**